rust/src/http2/http2.rs

```diff
@@ -0,0 +1,627 @@
+    transactions: Vec<HTTP2Transaction>,
+    progress: HTTP2ConnectionState,
+}
+
+impl HTTP2State {
+    pub fn new() -> Self {
+        Self {
+            tx_id: 0,
+            request_frame_size: 0,
+            response_frame_size: 0,
+            transactions: Vec::new(),
+            progress: HTTP2ConnectionState::Http2StateInit,
+        }
+    }
+
+    fn set_event(&mut self, event: HTTP2Event) {
+        let len = self.transactions.len();
+        if len == 0 {
+            return;
+        }
+        let tx = &mut self.transactions[len - 1];
+        let ev = event as u8;
+        core::sc_app_layer_decoder_events_set_event_raw(&mut tx.events, ev);
+    }
+
+    // Free a transaction by ID.
+    fn free_tx(&mut self, tx_id: u64) {
+        let len = self.transactions.len();
+        let mut found = false;
+        let mut index = 0;
+        for i in 0..len {
+            let tx = &self.transactions[i];
+            if tx.tx_id == tx_id + 1 {
+                found = true;
+                index = i;
+                break;
+            }
+        }
+        if found {
+            self.transactions.remove(index);
+        }
+    }
+
+    pub fn get_tx(&mut self, tx_id: u64) -> Option<&HTTP2Transaction> {
+        for tx in &mut self.transactions {
+            if tx.tx_id == tx_id + 1 {
+                return Some(tx);
+            }
+        }
+        return None;
+    }
+
+    fn new_tx(&mut self) -> HTTP2Transaction {
+        let mut tx = HTTP2Transaction::new();
+        self.tx_id += 1;
+        tx.tx_id = self.tx_id;
+        return tx;
+    }
+
+    fn parse_ts(&mut self, mut input: &[u8]) -> AppLayerResult {
+        //first consume frame bytes
+        let il = input.len();
+        if self.request_frame_size > 0 {
+            let ilen = input.len() as u32;
+            if self.request_frame_size >= ilen {
+                self.request_frame_size -= ilen;
+                return AppLayerResult::ok();
+            } else {
+                let start = self.request_frame_size as usize;
+                input = &input[start..];
+                self.request_frame_size = 0;
+            }
+        }
+
+        //then parse all we can
+        while input.len() > 0 {
+            match parser::http2_parse_frame_header(input) {
+                Ok((rem, head)) => {
+                    //TODO handle transactions the right way
+                    let mut tx = self.new_tx();
+                    tx.ftype = Some(head.ftype);
+                    match head.ftype {
+                        parser::HTTP2FrameType::GOAWAY => {
+                            match parser::http2_parse_frame_goaway(rem) {
+                                Ok((_, goaway)) => {
+                                    tx.type_data = Some(HTTP2FrameTypeData::GOAWAY(goaway));
+                                }
+                                Err(nom::Err::Incomplete(_)) => {
+                                    return AppLayerResult::incomplete(
```

Comment:
This seems the right way to do so far :
- parse frame header
- keep state of remaining frame bytes
- use tcp buffering if we need to parse deeper

Should be `HTTP2_FRAME_HEADER_LEN + 4`

Reply:
`nom::Err::Incomplete(_)` should tell you how many more bytes it needs. Can you use that too?

Reply:
Ok.
Should we try to be smarter : when we have a composite type, u16 and u32 for instance, we know we will need 6 bytes, even if nom tells us we first need 2 bytes ?